simpleviewinc/keg-cli
# Tap publish task

**Ticket**: [ZEN-503](https://jira.simpleviewtools.com/browse/ZEN-503)

## Notes
> This works with any repo / folder that is linked as a tap to the `keg-cli`
>  * Which means, literally any repo that is linked, and can be published to NPM will work
>  * For example I'm working on [this package](https://github.com/lancetipton/parkin)
>    * It's not a tap in anyway, but I linked it to the keg-cli with `keg tap link parkin`
>    * Then I ran `keg parkin publish` to publish it to npm. Here's a video example
>       * [Video of tap publish task](https://slack-files.com/TS0KQJ4UW-F01MZMSBUAX-861a624852)

> I plan to add a `getTapConfig` utility function which will get the `tap config` of a linked tap
> * This will allow taps to define their own publish config within their own `tap.json` config file
> * This also means they don't have to be defined in the `globalConfig`
>   * Currently they are not defined in the `globalConfig`, but they also can't be customized either
> * This would allow taps to be mono-repos _( keg-herkin )_, with multiple packages to publish
>   * For now **only the root package** of a tap can be published, which covers most use cases
>   * I have added some code for this, but it's currently not being used until the `getTapConfig` utility is added


## Context
* Currently the publish task only works for keg-hub repos.
* It should work for all linked taps

## Goal
* Allow publishing to NPM for all linked taps

## Updates
* `src/__mocks__/helpers/globalConfig.js`
  * Updates for testing publishing
* `src/tasks/hub/publish.js`
  * Extracted methods to allow them to be reused
* `src/tasks/tap/*`
  * Added tap publish task
* `src/utils/hub/getHubRepos.js`
  * Updated to allow finding tap repos based on the linked tap name
  * Has some code for publishing tap mono-repos, but not currently being used
* `src/utils/publish/__tests__/*`
  * Added tests for publish util methods
* `src/utils/publish/*`
  * Extracted methods from `src/tasks/hub/publish.js` so they can be reused in the `tap publish` task

## Testing

### Setup
* **Please** always use **patch** as the version update
* **Please** always add the **--dry** option when running the tasks
  * This ensure the changes are not published
  * But in the event you don't, hopefully you used **patch** as the version
  * This way we don't push unneeded version updates

### Test 1
* Pull this pr and run the tests => `keg cli && keg pr 34 && keg cli test`
  * Ensure all tests pass

### Test 2
* Run the `keg hub publish --dry` task for any of the keg-hub repos
  * Make sure you **add** `--dry` option when calling the task
  * For example `keg hub publish --context retheme --dry`
  * Ensure the publish task does not fail from a code error
    * If it fails for another reason, i.e. authentication, that's on you

### Test 3
* Run the `keg tap publish --dry` task for any linked tap
  * Again, make sure you **add** `--dry` option when calling the task
  * For example `keg evf publish --dry`
  * Ensure the publish task does not fail from a code error
    * If it fails for another reason, i.e. authentication, that's on you
